src/Microsoft.AspNetCore.SignalR.Core/SerializedHubMessage.cs

```diff
@@ -35,38 +38,84 @@ public SerializedHubMessage(HubMessage message)
 
         public ReadOnlyMemory<byte> GetSerializedMessage(IHubProtocol protocol)
         {
-            if (!TryGetCached(protocol.Name, out var serialized))
+            // Double-check locking!
+            if (!TryGetCachedFast(protocol.Name, out var serialized))
             {
-                if (Message == null)
+                lock (_lock)
                 {
-                    throw new InvalidOperationException(
-                        "This message was received from another server that did not have the requested protocol available.");
-                }
+                    if (!TryGetCached(protocol.Name, out serialized))
+                    {
+                        if (Message == null)
+                        {
+                            throw new InvalidOperationException(
+                                "This message was received from another server that did not have the requested protocol available.");
+                        }
 
-                serialized = protocol.GetMessageBytes(Message);
-                SetCache(protocol.Name, serialized);
+                        serialized = protocol.GetMessageBytes(Message);
+                        SetCache(protocol.Name, serialized);
+                    }
+                }
             }
 
             return serialized;
         }
 
+        // Used for unit testing.
+        internal IEnumerable<SerializedMessage> GetAllSerializations()
+        {
+            if (_count < 1)
+            {
+                yield break;
+            }
+
+            yield return _cachedItem1;
+
+            if (_count < 2)
+            {
+                yield break;
+            }
+
+            yield return _cachedItem2;
+
+            if (_count < 3)
+            {
+                yield break;
+            }
+
+            foreach (var item in _cachedItems)
+            {
+                yield return item;
+            }
+        }
+
         private void SetCache(string protocolName, ReadOnlyMemory<byte> serialized)
         {
+            // We set the fields before moving on to the list, if we need it to hold more than 2 items.
+            // In order to prevent "shearing" (where some of the fields of the struct are set by one thread,
+            // while another thread is reading the struct), we atomically increment a counter to ensure
+            // consumers know which structs have been fully assigned.
+            
             if (_cachedItem1.ProtocolName == null)
             {
                 _cachedItem1 = new SerializedMessage(protocolName, serialized);
+                Interlocked.Increment(ref _count);
```

Comment:
Is Interlocked.Increment required? SetCache is running under a lock so only one thread will access it at a time.

Reply:
The read isn't under the lock though. But I guess that's OK too now that I think about it. So yeah, Interlocked not required :)